src/tiledb/cloud/vcf/vcf_toolbox/annotate.py

```diff
@@ -0,0 +1,227 @@
+    reorder: Optional[Sequence[str]] = [
+        "sample_name",
+        "contig",
+        "pos_start",
+        "ref",
+        "alt",
+    ],
+    rename: Optional[Mapping[str, str]] = {"sample_name": "sample"},
+    verbose: bool = False,
+) -> pd.DataFrame:
+    """
+    Annotate a VCF DataFrame with annotations from a TileDB array.
+
+    Parameters
+    ----------
+    vcf_df
+        The input VCF DataFrame.
+    ann_uri
+        The URI of the annotation array.
+    ann_regions
+        The regions to query. All regions must be in the same chromosome/contig.
+    ann_attrs, optional
+        The attributes to query, by default None which queries all attributes.
+    vcf_filter, optional
+        A pandas filter to apply to the VCF DataFrame before annotation, by default None
+    split_multiallelic, optional
+        Split multiallelic variants into separate rows, by default True
+    add_zygosity, optional
+        Add a zygosity column to the DataFrame, by default False
+    reorder, optional
+        List of columns to reorder, by default
+            ["sample_name", "contig", "pos_start", "ref", "alt"]
+    rename, optional
+        Dict of columns to rename, by default
+            {"sample_name": "sample"}
+    verbose, optional
+        Enable verbose logging, by default False
+
+    Returns
+    -------
+        The annotated VCF DataFrame.
+    """
+
+    if isinstance(ann_attrs, str):
+        ann_attrs = [ann_attrs]
+
+    # Add attributes required for join
+    if ann_attrs is not None:
+        for attr in ["ref", "alt"]:
+            if attr not in ann_attrs:
+                ann_attrs.append(attr)
+
+    level = logging.DEBUG if verbose else logging.INFO
+    logger = get_logger(level)
+
+    def log_event(message, t_prev=None):
+        if t_prev:
+            message += f" {time.time() - t_prev:.3f} sec"
+        logger.debug(message)
+        return time.time()
+
+    t_prev = log_event("start annotation")
+    t_start = t_prev
+    mem_start = max_memory_usage()
+
+    # Split regions
+    contig, regions = split_regions(ann_regions)
+
+    # Start annotation query in a separate thread
+    def read_ann():
+        with tiledb.open(ann_uri) as A:
+            df = A.query(attrs=ann_attrs).df[contig, regions]
+        return df
+
+    executor = ThreadPoolExecutor(1)
+    ann_future = executor.submit(read_ann)
+
+    # Filter the VCF data
+    if vcf_filter:
+        vcf_df = vcf_df.query(vcf_filter).reset_index(drop=True)
+        t_prev = log_event("filter", t_prev)
+
+    # Return if the filtered VCF data is empty
+    if len(vcf_df) == 0:
+        return vcf_df
+
+    # Split alleles into ref and alt
+    vcf_df["ref"] = vcf_df["alleles"].str[0]
+    vcf_df["alt"] = vcf_df["alleles"].str[1:]
+    vcf_df = vcf_df.drop("alleles", axis=1)
+    t_prev = log_event("split ref/alt", t_prev)
+
+    if split_multiallelic:
+        # Split multiallelic variants
+        vcf_df = vcf_df.explode("alt")
```

Comment:
do the allele frequencies from `info_TILEDB_IAF` get duplicated to every allele here? otherwise how would we know which af goes with which allele? in the do_query routine we have:
```
frequencies = pandas.DataFrame(vcf_df["info_TILEDB_IAF"].to_list())
vcf_df["af"] = frequencies.drop([0], axis=1).values.tolist()
vcf_df = vcf_df.explode(["alt", "af"])
```

Reply:
Good point. Adding code to explode `info_TILEDB_IAF` (if present) and `alt` as you described.